### ISSUE TITLE
outlier analysis

### ISSUE THREAD

**Issue by ehb54** [ehb54/ustfpp#6]
[Sheet on a 64-64 PCSA](https://docs.google.com/spreadsheets/d/1CiEne1yFAUmxcA5omMlDXC8Yra5H2YT-hg2aNHiSagQ/edit?gid=2134528525#gid=2134528525)

On the validation sheet, the first 3 show an obvious issue, huge (outside of training range) values in columns X,Y

Predictions for test values outside the training range are almost sure to be terrible.
We are going to need to store the min/max for each data column for the training data and either cap the validation values to this range or reject.
For our actual performance prediction, this will be important. We can fall-back to our existing heuristic values for these cases. This is for all methods. 

In particular, for PCSA and possibly other methods, it could be interesting to eliminate some columns, e.g. do we really need the min/max values, are they important?

Probably don't have time for this later investgation, but a nice bit for the discussion.

**Comment by ehb54** [comment]
Hopefully, can look at some 2DSA & GA runs tomorrow for more outlier checking.
I suspect managing this issue will cure some of the issues we have been seeing.

**Comment by ehb54** [comment]
Looked at 2DSA runs and saw the same issue as we had with PCSA (validation loss uniformly much lower than training loss). Rerunning the 2DSA with a [randomized csv.](https://github.com/ehb54/ustfpp/commit/8ef3f803b560008d29a856a79002c1b8ce96ccad)

**Comment by ehb54** [comment]
Started to look at the GA runs, but found no loss curve pngs.
Will need to restart, but all GPUs on demeler9 are busy now with 2DSA.
Will try on gputests1 (ugh, new docker image needed)

**Comment by ehb54** [comment]
gputest1 gpu 100% on GA jobs (5 simultaneous)

**Comment by ehb54** [comment]
Some notes on gputest1 tf.
starting the container
```
docker run -it --rm --gpus all -v ~/ustfpp:/root/work --device /dev/nvidia0:/dev/nvidia0 --device /dev/nvidiactl:/dev/nvidiactl --device  /dev/nvidia-caps:/dev/nvidia-caps --device  /dev/nvidia-modeset:/dev/nvidia-modeset --device  /dev/nvidia-uvm:/dev/nvidia-uvm --device  /dev/nvidia-uvm-tools:/dev/nvidia-uvm-tools --runtime=nvidia tf18:2 bash
```
Once the container is started, need at the cli
```
export CUDNN_PATH=$(dirname $(python -c "import nvidia.cudnn;print(nvidia.cudnn.__file__)"))
export LD_LIBRARY_PATH=${CUDNN_PATH}/lib
```
to get tf to see the gpu

quickie to check if gpu seen
```
print("Num GPUs Available: ", len(tf.config.list_physical_devices('GPU')))
```
---
### notes about what was done:
---
various packages already pip installed (perhaps some others)
```
keras
seaborn
matplotlib
scikit-learn
```
used `docker pull tensorflow/tensorflow:latest-gpu`
as it at first complained about wrong nvidia cuda version, had to use --env NVIDIA_DISABLE_REQUIRE=1 on the docker run line, but after I committed the container, it was no longer needed

also had to [install cuDNN](https://docs.nvidia.com/deeplearning/cudnn/installation/latest/linux.html) in the container
iirc `apt install libcudnn9-cuda-12 && pip install nvidia-cudnn-cu12`

**Comment by ehb54** [comment]
Update 
Spent some time with 2DSA outlier analysis.
Created a new filtered set for 2DSA, rerunning.
As with PCSA, majority excellent predictions, but some were terrible.
I'll see what these results bring in the morning.
Also created a [utility](https://github.com/ehb54/ustfpp/blob/ehb-dev2/util/collect_results.php) script to collect best data for easy downloading ;)
Sorry, it's in php, but was fastest for me to do.

**Comment by ehb54** [comment]
@aaron-auc, forget the request above, I see the column in the data already.

**Comment by ehb54** [comment]
@aaron-auc 
Have you done anything in the preprocessing for outlier analysis?
Here's a simple [writeup](https://www.analyticsvidhya.com/blog/2021/05/detecting-and-treating-outliers-treating-the-odd-one-out/), likely better ones out there.
Something better than min/max on the data. Not sure what method is best, but my initial thought is to do some outlier analysis on each feature and see if our prediction bad values (worst 0.5 %) are in any way feature outliers.
I'll work on this later today, but wanted to check if you had done anything or had anything to suggest...

**Comment by aaron-auc** [comment]
I hadn't done anything with outlier analysis as part of the preprocessing step. 

If you just want a visual then a box plot of the features might give you an idea of how extreme an outlier is for each feature. It might also depend on how the feature data is distributed. If it's a normal distribution you might remove the ends or leave it alone. If it's heavily skewed then you might pick something at the tail. It's like when looking at the data yesterday, if 99.9999% of the data is between 0-800 and then you have these 15,000,000 values then that seems like a true outlier that might decrease the accuracy.

I updated my code last night for the feature analysis to use random forest, lasso, and xgboost and to run it as a Monte Carlo to select features. I'm working on getting that to run today to reduce the feature set and test today.

**Comment by ehb54** [comment]
Some positive news
```
$ ~/srv/ustfpp/util/find_outliers.php --file 2dsa-filtered-randomized-nonneg.csv --z-score 3

================================================================================
summary
================================================================================
total data rows 225853
rows accepted 201619 (89.27 %)
rows rejected 24234 (10.73 %)
```

Rejected rows with any column outsize of +/-3 SD from Mean.

Spot checked the top 3 worst validation absolute error & they *were* in the rejection set :)
Still need to create a utility to thoroughly check and then balance the z-score rejection and the absolute error (i.e. get acceptable predictions for the maximum % of jobs). Perhaps a nice additional plot.

This utility also creates a json file with the column names and values which can be consumed to accept or reject during performance prediction

Also, if we have time, after I'm done with my 'thoroughly check' analysis, if training with or without the outliers performs better/worse or the same.

**Comment by aaron-auc** [comment]
The preprocessor can now remove min/max values as part of processing. I made everything configurable and added an md document to explain how to configure the preprocessor. 

I started with our full dataset, processed it, and then kicked off the feature analysis. So we do have a repeatable, configurable process up to that point.

**Comment by ehb54** [comment]
Basic maximum absolute error analysis pipeline working.
Will show when we next meet. 
Should look at training on outlier rejected training data, but at what z-score?
Feature selected training too with and without preprocess outlier rejection.

**Comment by ehb54** [comment]
Manuscript submitted
